File: day15/day15.cpp
#include <array>
#include <string>
#include <fstream>
#include <iostream>

int main()
{
  std::ifstream in { "input.txt" };
  std::string input { std::istreambuf_iterator<char> {in}, std::istreambuf_iterator<char>{}};
  std::array<int, 7> nums;
  std::size_t pos = 0, proc = 0;
  int i = 0;
  while (pos < input.size()) {
    nums[i++] = std::stoi(input.substr(pos), &proc);
    pos += proc + 1;
  }
  auto *prevs = new std::array<int, 30000000>{};
  for (int i = 1; i <= nums.size(); ++i) {
    (*prevs)[nums[i-1]] = i;
  }
  int last = nums.back();
  for (int turn = nums.size(); turn < 30000000; ++turn) {
    int n = (*prevs)[last] ? turn - (*prevs)[last] : 0;
    (*prevs)[last] = turn;
    last = n;
    if (turn == 2019) {
      std::cout << last << '\n';  // first answer
    }
  }
  std::cout << last << '\n';  // second answer
  return 0;
}

File: day15/day15.py
#!/usr/bin/env python3

# Running time (on my laptop) of today's solutions:
# day15.py with CPython3.7: ~5.3s
# day15.py with pypy3: ~1.2s
# day15.pyx compiled with Cython 0.29 in CPython3.7 console: ~3.7s
# day15.cpp compiled with g++ 7.5.0 with -Ofast: 0.8s

def run():
    with open('input.txt') as fh:
        nums = [int(n) for n in fh.read().split(',')]

    prevs = [0] * 30000000
    for i, n in enumerate(nums, 1):
        prevs[n] = i
    last = nums[-1]
    for turn in range(len(nums), 30000000):
        prevs[last], last = turn, turn - prevs[last] if prevs[last] else 0
        if turn == 2019:
            print(last)  # first answer
    print(last)  # second answer

if __name__ == '__main__':
    run()
